addons/sale/models/sale_order_line.py

```diff
@@ -5,17 +5,247 @@
 
 from odoo import api, fields, models, _
 from odoo.exceptions import UserError
-from odoo.tools.misc import get_lang
 from odoo.osv import expression
 from odoo.tools import float_is_zero, float_compare, float_round
 
 
 class SaleOrderLine(models.Model):
     _name = 'sale.order.line'
-    _description = 'Sales Order Line'
+    _description = "Sales Order Line"
     _order = 'order_id, sequence, id'
     _check_company_auto = True
 
+    # Fields are ordered according by tech & business logics
+    # and computed fields are defined after their dependencies.
+    # This reduces execution stacks depth when precomputing fields
+    # on record creation (and is also a good ordering logic imho)
+
+    order_id = fields.Many2one(
+        comodel_name='sale.order',
+        string="Order Reference",
+        required=True, ondelete='cascade', index=True, copy=False)
+    sequence = fields.Integer(string="Sequence", default=10)
+
+    # Order-related fields
+    company_id = fields.Many2one(
+        related='order_id.company_id',
+        store=True, index=True, precompute=True)
+    currency_id = fields.Many2one(
+        related='order_id.currency_id',
+        depends=['order_id.currency_id'],
+        store=True, precompute=True)
+    order_partner_id = fields.Many2one(
+        related='order_id.partner_id',
+        string="Customer",
+        store=True, precompute=True)
+    salesman_id = fields.Many2one(
+        related='order_id.user_id',
+        string="Salesperson",
+        store=True, precompute=True)
+    state = fields.Selection(
+        related='order_id.state',
+        string="Order Status",
+        copy=False, store=True, precompute=True)
+
+    # Fields specifying custom line logic
+    display_type = fields.Selection(
+        selection=[
+            ('line_section', "Section"),
+            ('line_note', "Note"),
+        ],
+        help="Technical field for UX purpose.",
+        default=False)
+    is_downpayment = fields.Boolean(
+        string="Is a down payment",
+        help="Down payments are made when creating invoices from a sales order."
+            " They are not copied when duplicating a sales order.")
+    is_expense = fields.Boolean(
+        string="Is expense",
+        help="Is true if the sales order line comes from an expense or a vendor bills")
+
+    # Generic configuration fields
+    product_id = fields.Many2one(
+        comodel_name='product.product',
+        string="Product",
+        change_default=True, ondelete='restrict', check_company=True,
+        domain="[('sale_ok', '=', True), '|', ('company_id', '=', False), ('company_id', '=', company_id)]")
+    product_template_id = fields.Many2one(
+        string="Product Template",
+        related='product_id.product_tmpl_id',
+        domain=[('sale_ok', '=', True)])
+    product_uom_category_id = fields.Many2one(related='product_id.uom_id.category_id')
```

Comment:
add `depends=['product_id']` to prevent big triger @rco-odoo

Reply:
Useless, the field is not stored and only computed when the record is shown.